### PR TITLE
[BEAM-7824] Sets a default environment for Dataflow runner

### PR DESCRIPTION
**Pull request by chamikaramj**
* Sets a default environment for Dataflow runner
* Also sets the default environment and environments received from External transform to runner API proto. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by chamikaramj** [comment]
R: @ihji 

CC: @robertwb @angoenka

**Comment by chamikaramj** [comment]
Run Python PostCommit

**Comment by chamikaramj** [comment]
Run Python 2 PostCommit

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 1]
```diff
@@ -952,7 +969,8 @@ def _get_required_container_version(job_type=None):
         current version of the SDK.
     """
   if 'dev' in beam_version.__version__:
-    if job_type == 'FNAPI_BATCH' or job_type == 'FNAPI_STREAMING':
+    if (job_type == JOB_TYPE_PYTHON_BATCH or
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: python-batch or python-fnapi-streaming?

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 2]
```diff
@@ -927,7 +944,7 @@ def get_default_container_image_for_current_sdk(job_type):
                     % str(sys.version_info[0:2]))
 
   # TODO(tvalentyn): Use enumerated type instead of strings for job types.
-  if job_type == 'FNAPI_BATCH' or job_type == 'FNAPI_STREAMING':
+  if job_type == JOB_TYPE_PYTHON_BATCH or job_type == JOB_TYPE_FNAPI_STREAMING:
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: _FNAPI_

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 3]
```diff
@@ -379,9 +379,18 @@ def run_pipeline(self, pipeline, options):
     if apiclient._use_sdf_bounded_source(options):
       pipeline.replace_all(DataflowRunner._SDF_PTRANSFORM_OVERRIDES)
 
+    from apache_beam.portability.api import beam_runner_api_pb2
+    default_container_image = (
+        apiclient.get_default_container_image_for_current_sdk(
+            apiclient.get_job_type(options)))
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: This value is only used for FnApi. Additionally, FnApi doesn't have different containers for streaming vs. batch. It's too bad we have to go through the (irrelevant for portability) job type to get this.

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py (hunk 1 above)
Comment (chamikaramj) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: Reverted.

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py (hunk 2 above)
Comment (chamikaramj) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: Reverted.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 3 above)
Comment (chamikaramj) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Updated function to take a bool (use_fnapi) instead of job type.

**Comment by chamikaramj** [comment]
Thanks Robert. PTAL.

File: sdks/python/apache_beam/transforms/core.py [hunk 4]
```diff
@@ -308,14 +308,19 @@ class RunnerAPIPTransformHolder(PTransform):
...
+  def __init__(self, proto, context):
     self._proto = proto
+    self._context = context
 
   def proto(self):
     """Runner API payload for a `PTransform`"""
     return self._proto
 
   def to_runner_api(self, context, has_parts=False):
+    id_to_proto_map = self._context.environments.get_id_to_proto_map()
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: This seems a bit fragile in that we're assuming that if the ids match, the protos match (which could be bad for auto-generated names like env0). Could you add a check for this. 

It also seems that we're copying too much (every environment from the context, not just the one(s) referenced from this proto), but perhaps there's no good way to get around that. Could you at least add a TODO referencing the JIRA about making environment a top-level attribute of transforms?

Comment (chamikaramj) [reply sdks/python/apache_beam/transforms/core.py]: Added a check for environments.

Created https://issues.apache.org/jira/browse/BEAM-7850 for making Environment a top level attribute of PTransform.

**Comment by chamikaramj** [comment]
Run Python_PVR_Flink PreCommit

**Comment by chamikaramj** [comment]
Flink failure seems to be unrelated.

**Pull request merged**